### DISCUSSION TITLE
OpenTelemetry HTTP header Propogation

### DISCUSSION THREAD

**Discussion by linktech1** [quarkusio/quarkus#36513] [Q&A]
Is there not a way to auto propagate the parentId and traceId in the RestEasy reactive client headers? We can make it work using ClientRequestFilter or org.eclipse.microprofile.rest.client.propagateHeaders=traceparent,tracestate
 but I would rather not have to add any of these to propagate the W3C Trace Context headers.

[a comment by an automated account is not shown]

**Reply by brunobat** [reply]
`quarkus-rest-client-reactive-jackson` has OTel support and should have context propagation. Do you have a code example that doesn't work?

**Comment by linktech1** [comment]
I took the Otel example and added rest client. Without using org.eclipse.microprofile.rest.client.propagateHeaders=traceparent,tracestate
No Otel headers are sent. 

=============================================================
package org.acme.opentelemetry;

import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.core.Context;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.UriInfo;

import org.eclipse.microprofile.rest.client.RestClientBuilder;
import org.eclipse.microprofile.rest.client.inject.RestClient;
import org.jboss.logging.Logger;

import io.smallrye.mutiny.Uni;
import io.smallrye.mutiny.tuples.Tuple3;

@Path("/")
public class TracedResource {

    private static final Logger LOG = Logger.getLogger(TracedResource.class);

    @Context
    UriInfo uriInfo;

    @RestClient
    Countries countries;
    
    @GET
    @Path("/hello")
    @Produces(MediaType.TEXT_PLAIN)
    public String hello() {
        LOG.info("hello");
        return "hello";
    }
    
    @GET
    @Path("/all")
    @Produces(MediaType.APPLICATION_JSON)
    public String all() {
        LOG.info("all");
        countries.all();
        countries.all();
        return countries.all();
    }

    @GET
    @Path("/allAsync")
    @Produces(MediaType.APPLICATION_JSON)
    public Uni<String> allAsync()
    {
    	Uni<Tuple3<String, String, String>> tuple = 
    		Uni
    			.combine()
    			.all()
    			.unis(
    				countries.allAsync(), 
    				countries.allAsync(),
    				countries.allAsync()) .asTuple();
        LOG.info("allAsync");
    	return tuple.chain(aTuple -> 
    		getTupleContents(aTuple.getItem1()));
    	
    }
    
    public Uni<String> getTupleContents(String tupleContents)
    {
    	return Uni.createFrom().item(tupleContents);
    }

    @GET
    @Path("/chain")
    @Produces(MediaType.TEXT_PLAIN)
    public String chain() {
        ResourceClient resourceClient = RestClientBuilder.newBuilder()
                .baseUri(uriInfo.getBaseUri())
                .register((MyFilter.class))
                .build(ResourceClient.class);
        return "chain -> " + resourceClient.hello();
    }
}

================================================================

package org.acme.opentelemetry;

import org.eclipse.microprofile.rest.client.annotation.RegisterClientHeaders;
import org.eclipse.microprofile.rest.client.annotation.RegisterProvider;
import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;

import io.smallrye.mutiny.Uni;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.core.MediaType;

@Path("")
@RegisterRestClient(configKey="countries")
@Produces("application/json")
@RegisterClientHeaders
@RegisterProvider(MyFilter.class)
public interface Countries {
	   @GET
	    @Path("/all")
	    @Produces(MediaType.APPLICATION_JSON)
	    String all();
	   
	   @GET
	    @Path("/all")
	    @Produces(MediaType.APPLICATION_JSON)
	   Uni<String> allAsync();
}

===========================================================

quarkus.application.name=myservice
quarkus.otel.exporter.otlp.traces.endpoint=http://localhost:4317
quarkus.otel.exporter.otlp.traces.headers=Authorization=Bearer my_secret
quarkus.log.console.format=%d{HH:mm:ss} %-5p traceId=%X{traceId}, parentId=%X{parentId}, spanId=%X{spanId}, sampled=%X{sampled} [%c{2.}] (%t) %s%e%n

##org.eclipse.microprofile.rest.client.propagateHeaders=traceparent,tracestate

quarkus.rest-client.countries.url=https://restcountries.com/v3.1/

quarkus.log.min-level=ALL

quarkus.http.access-log.enabled=true
quarkus.http.access-log.pattern=%h %l %u %t "%r" %s %b\n%{ALL_REQUEST_HEADERS}
##quarkus.http.access-log.pattern=%{REMOTE_HOST} %l %{REMOTE_USER} %{DATE_TIME} "%{REQUEST_LINE}" %{RESPONSE_CODE} %b %{RESPONSE_TIME} %T

quarkus.http.access-log.log-to-file=true
quarkus.log.category."org.apache.http".level=DEBUG

quarkus.rest-client.logging.scope=all
quarkus.rest-client.logging.body-limit=1024
quarkus.log.category."org.jboss.resteasy.reactive.client.logging".level=DEBUG

===========================================================

<?xml version="1.0"?>
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <modelVersion>4.0.0</modelVersion>
  <groupId>org.acme</groupId>
  <artifactId>opentelemetry-quickstart</artifactId>
  <version>1.0.0-SNAPSHOT</version>
  <properties>
    <quarkus.platform.artifact-id>quarkus-bom</quarkus.platform.artifact-id>
    <quarkus.platform.group-id>io.quarkus.platform</quarkus.platform.group-id>
    <quarkus.platform.version>3.4.3</quarkus.platform.version>
    <surefire-plugin.version>3.1.2</surefire-plugin.version>
    <maven.compiler.source>17</maven.compiler.source>
    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    <maven.compiler.target>17</maven.compiler.target>
  </properties>
  <dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>${quarkus.platform.group-id}</groupId>
        <artifactId>${quarkus.platform.artifact-id}</artifactId>
        <version>${quarkus.platform.version}</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
    </dependencies>
  </dependencyManagement>
  <dependencies>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-resteasy-reactive</artifactId>
    </dependency>
    
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-resteasy-reactive-jackson</artifactId>
    </dependency>

    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-junit5</artifactId>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>io.rest-assured</groupId>
      <artifactId>rest-assured</artifactId>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-opentelemetry</artifactId>
    </dependency>
    <dependency>
    	<groupId>io.quarkus</groupId>
    	<artifactId>quarkus-micrometer-registry-prometheus</artifactId>
	</dependency>

    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-rest-client-reactive</artifactId>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-rest-client-reactive-jackson</artifactId>
    </dependency>

    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-info</artifactId>
    </dependency>

  </dependencies>
  <build>
    <plugins>
      <plugin>
        <artifactId>maven-surefire-plugin</artifactId>
        <version>${surefire-plugin.version}</version>
        <configuration>
          <systemPropertyVariables>
            <java.util.logging.manager>org.jboss.logmanager.LogManager</java.util.logging.manager>
            <maven.home>${maven.home}</maven.home>
          </systemPropertyVariables>
        </configuration>
      </plugin>
      <plugin>
        <groupId>${quarkus.platform.group-id}</groupId>
        <artifactId>quarkus-maven-plugin</artifactId>
        <version>${quarkus.platform.version}</version>
        <executions>
          <execution>
            <goals>
              <goal>build</goal>
            </goals>
          </execution>
        </executions>
      </plugin>
    </plugins>
  </build>
  <profiles>
    <profile>
      <id>native</id>
      <activation>
        <property>
          <name>native</name>
        </property>
      </activation>
      <properties>
        <quarkus.package.type>native</quarkus.package.type>
      </properties>
      <build>
        <plugins>
          <plugin>
            <artifactId>maven-failsafe-plugin</artifactId>
            <version>${surefire-plugin.version}</version>
            <executions>
              <execution>
                <goals>
                  <goal>integration-test</goal>
                  <goal>verify</goal>
                </goals>
                <configuration>
                  <systemPropertyVariables>
                    <native.image.path>${project.build.directory}/${project.build.finalName}-runner</native.image.path>
                    <java.util.logging.manager>org.jboss.logmanager.LogManager</java.util.logging.manager>
                    <maven.home>${maven.home}</maven.home>
                  </systemPropertyVariables>
                </configuration>
              </execution>
            </executions>
          </plugin>
        </plugins>
      </build>
    </profile>
  </profiles>
</project>

==================================================================

**Reply by brunobat** [reply]
I'd say that's a bug, can you please file an issue for it?
